zilliqa/src/contracts/deposit.sol

```diff
@@ -138,61 +207,238 @@ contract Deposit {
+            _stakerKeys[msg.sender] = blsPubKey;
         }
 
-        _stakersMap[blsPubKey].balance += msg.value;
-        totalStake += msg.value;
+        queueNewDeposit(blsPubKey, msg.value);
+    }
 
-        if (_stakersMap[blsPubKey].balance < _minimumStake) {
-            revert("stake less than minimum stake");
+    function depositTopup() public payable isStaker() {
+        bytes memory blsPubKey = _stakerKeys[msg.sender];
+        queueNewDeposit(blsPubKey, msg.value);
+    }
+
+    function queueNewDeposit(bytes memory blsPubKey, uint256 amount) private {
+        Value storage newDeposit = pendingQueues[currentQueueIndex()].pushBack();
+        newDeposit.key = blsPubKey;
+        newDeposit.amount = amount;
+    }
+
+    function unstake(
+        uint256 amount
+    ) public isStaker() {
+        bytes storage blsPubKey = _stakerKeys[msg.sender];
+
+        uint256 stakedBalance = _stakersMap[blsPubKey].stakedBalance - _stakersMap[blsPubKey].unstakingBalance;
+        if (amount > stakedBalance) {
+            revert("Cannot unstake more than current validator staked balance.");
         }
+        _stakersMap[blsPubKey].unstakingBalance += amount;
+
+        Value storage newUnstaking = unstakingQueues[currentQueueIndex()].pushBack();
+        newUnstaking.key = blsPubKey;
+        newUnstaking.amount = amount;
+    }
 
-        _stakersMap[blsPubKey].rewardAddress = rewardAddress;
-        _stakersMap[blsPubKey].peerId = peerId;
+    function withdrawFunds(
+        uint256 amount
+    ) public isStaker() {
+        bytes storage blsPubKey = _stakerKeys[msg.sender];
+        if (amount > _stakersMap[blsPubKey].freeBalance) {
+            revert("Attempting to withdraw more than the available free balance.");
+        }
+        _stakersMap[blsPubKey].freeBalance -= amount;
+        (bool sent, bytes memory _data) = payable(_stakersMap[blsPubKey].controlAddress).call{value: amount}("");
+        require(sent, "Unable to transfer balance");
+    }
+
+    function suspend() public isStaker() {
+        bytes storage blsPubKey = _stakerKeys[msg.sender];
+        if (_stakersMap[blsPubKey].status == Status.Slashed) {
+            revert("A slashed validator cannot self-suspend.");
+        }
+
+        // Note: we immediately set the validator status to suspended here.
+        // This is very generous to the validators: they can simply pause operation at any point, with no penalty.
+        // An alternative would be to keep the validator Active. Then it would be removed from consensus at the
+        // next epoch boundary (a fast-track compared to simply withdrawing all stake, which would take up to 2 epochs),
+        // but would remain liable for liveness until then.
+        _stakersMap[blsPubKey].status = Status.Suspended;
+
+        Value storage newSuspension = kickQueue.pushBack();
+        newSuspension.key = blsPubKey;
+        // no associated amount
+    }
+
+    function eject(
```

Comment:
this is the successor of `tempRemoveStaker()` and will be removed once penalties are fully implemented, right?